<launch>

    <arg name="namespace" default="LM_Mapper"/>
    <arg name="launch_rviz" default="true" />
    <arg name="process_pcl_at_startup" default="true" />
    <arg name="inertial_frame_id" default="/map" />
    <arg name="depth_cam_prefix" default="/" />
    <arg name="depth_cam_suffix" default="/camera/depth/compressed_points" />

    <group ns="$(arg namespace)">
        <node pkg="mapper"
            type="mapper"
            name="mapper_node"
            output="screen">

            <!-- Namespace -->
            <param name="namespace" value="$(arg namespace)"/>

            <!-- Set of depth camera topics -->
            <param name="depth_cam_prefix" value="$(arg depth_cam_prefix)"/>
            <rosparam param="depth_cam_names"> ["Sauron"] </rosparam>
            <param name="depth_cam_suffix" value="$(arg depth_cam_suffix)"/>

            <!-- tf frame ID for the cameras above -->
            <rosparam param="cam_frame_id"> ["/Sauron/camera_depth_optical_frame"] </rosparam>

            <!-- Inertial frame name (used for getting tf tree information) -->
            <param name="inertial_frame_id" value="$(arg inertial_frame_id)"/>

            <!-- Set parameters for the mapper -->
            <param name="map_resolution" value="0.10"/>          <!-- meters -->
            <param name="max_range" value="3"/>                 <!-- meters -->
            <param name="min_range" value="0.5"/>              <!-- meters -->

            <!-- Time to forget information (set as negative if infinity) -->
            <param name="memory_time" value="10.0"/>           <!-- seconds -->

            <!-- Fading memory thread update rate (Hz) -->
            <param name="fading_memory_update_rate" value="10"/><!-- Hz -->

            <!-- Map inflation parameters -->
            <param name="inflate_radius_xy" value="0.30"/>       <!-- meters -->
            <param name="inflate_radius_z" value="0.30"/>       <!-- meters -->

            <!-- Camera FOV (radians): map is only updated within this FOV -->
            <param name="cam_fov" value="0.8"/>             <!-- radians -->

            <!-- Camera's width divided by height (unitless) -->
            <param name="cam_aspect_ratio" value="1.3099"/>

            <!-- Map probabilities -->
            <param name="occupancy_threshold" value="0.5"/>     <!-- 0-1 -->
            <param name="probability_hit" value="0.7"/>         <!-- 0-1 -->
            <param name="probability_miss" value="0.25"/>       <!-- 0-1 -->
            <param name="clamping_threshold_min" value="0.1"/>  <!-- 0-1 -->
            <param name="clamping_threshold_max" value="0.9"/>  <!-- 0-1 -->

            <!-- Collision Checking parameters -->
            <param name="traj_compression_max_dev" value="0.01"/>     <!-- meters -->
            <param name="traj_compression_resolution" value="0.02"/>  <!-- meters -->
            <param name="collision_check_rate" value="0.01"/>         <!-- Hz -->

            <!-- Frequency at which tf listeners update tf -->
            <param name="tf_update_rate" value="50"/>            <!-- Hz -->

            <!-- Service names -->
            <param name="update_resolution" value="update_resolution"/>
            <param name="update_memory_time" value="update_memory_time"/>
            <param name="update_inflation_radius" value="update_inflation_radius"/>
            <param name="reset_map" value="reset_map"/>
            <param name="save_map" value="save_map"/>
            <param name="load_map" value="load_map"/>
            <param name="process_pcl" value="process_pcl"/>
            <param name="rrg_service" value="rrg"/>

            <!-- Publisher names -->
            <param name="obstacle_markers" value="obstacle_markers"/>
            <param name="free_space_markers" value="free_space_markers"/>
            <param name="inflated_obstacle_markers" value="inflated_obstacle_markers"/>
            <param name="inflated_free_space_markers" value="inflated_free_space_markers"/>
            <param name="frustum_markers" value="frustum_markers"/>
            <param name="discrete_trajectory_markers" value="discrete_trajectory_markers"/>
            <param name="collision_detection" value="collisions"/>
            <param name="graph_tree_marker_topic" value="graph_tree_path_planner"/>

            <!-- Path of mapper in computer: used for saving/loading maps -->
            <param name="local_path" value="$(find mapper)"/>

            <!-- Whether or not PCL data is processed at startup -->
            <!-- If chosen as false, pcl processing only starts after request through a service -->
            <param name="process_pcl_at_startup" value="$(arg process_pcl_at_startup)"/>

        </node>

    </group>

    <!-- Launch RVIZ -->
    <group if="$(arg launch_rviz)">
        <node pkg="rviz" type="rviz" name="rviz" args="-d $(find mapper)/cfg/mapper.rviz"/>
    </group>

</launch>